website/sales/api/v2/serializers/order.py

```diff
@@ -0,0 +1,192 @@
+        except ObjectDoesNotExist:
+            self.fail(
+                "does_not_exist", slug_name=self.slug_field, value=smart_str(data)
+            )
+        except (TypeError, ValueError):
+            self.fail("invalid")
+
+    def to_representation(self, obj):
+        return obj.product.name
+
+
+class OrderItemSerializer(serializers.ModelSerializer):
+    """Serializer for order items."""
+
+    class Meta:
+        model = OrderItem
+        fields = ("product", "amount", "total")
+        read_only_fields = ("total",)
+
+    product = ProductNameRelatedField("product")
+
+    total = serializers.DecimalField(
+        max_digits=6, decimal_places=2, min_value=0, read_only=True
+    )
+
+    def get_fields(self):
+        fields = super().get_fields()
+        request = self.context.get("request", None)
+        if request and request.user and request.user.has_perm("sales.custom_prices"):
+            fields["total"].read_only = False
+        return fields
+
+    def create(self, validated_data, **kwargs):
+        order = self.context["order"]
+        item = OrderItem.objects.create(order=order, **validated_data)
+        return item
+
+    def update(self, instance, validated_data, **kwargs):
+        order = self.context["order"]
+        instance.order = order
+        instance.total = None  # Always recalculate the total amount if updating using API (note the difference from the model that only recalculates if the total is None, to deal with historic data and allow for special discounts)
+        super().update(instance, validated_data)
+        return instance
+
+
+class OrderSerializer(serializers.ModelSerializer):
+    """Serializer for orders."""
+
+    class Meta:
+        model = Order
+        fields = (
+            "pk",
+            "shift",
+            "created_at",
+            "order_items",
+            "order_description",
+            "age_restricted",
+            "subtotal",
+            "discount",
+            "total_amount",
+            "num_items",
+            "payment",
+            "payer",
+            "payment_url",
+        )
+        read_only_fields = (
+            "pk",
+            "shift",
+            "created_at",
+            "discount",
+            "payment",
+            "total_amount",
+            "num_items",
+            "age_restricted",
+            "order_description",
+            "payment_url",
+        )
+
+    shift = serializers.PrimaryKeyRelatedField(read_only=True)
+
+    age_restricted = serializers.BooleanField(read_only=True)
+
+    order_items = OrderItemSerializer(many=True, required=False)
+
+    subtotal = serializers.DecimalField(
+        max_digits=6, decimal_places=2, min_value=0, read_only=True
+    )
+
+    discount = serializers.DecimalField(
+        max_digits=6, decimal_places=2, min_value=0, read_only=True
+    )
+
+    total_amount = serializers.DecimalField(
+        max_digits=6, decimal_places=2, min_value=0, read_only=True
+    )
```

Comment:
I do not entirely understand why these fields are not just part of the `read_only_fields` list. Please enlighten me :)

Reply:
I also see that some still have the read only attribute and are in that list as well?

Reply:
I see, I'm not consistent in this. I do need to specify it are decimal fields explicitly, because Django is not able to correctly derive the field type for these queryable properties on all database systems. The read only thing can be done more consistently. Will fix